File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum = Integer.MIN_VALUE; // Initialize maxSum with the smallest possible integer value
    
    public int maxPathSum(TreeNode root) {
        maxGain(root); // Start the DFS traversal
        return maxSum;
    }
    
    private int maxGain(TreeNode node) {
        if (node == null) return 0; // Base case: null node contributes 0 to the path sum
        
        // Recursively calculate the maximum gain from the left and right subtrees
        int leftGain = Math.max(maxGain(node.left), 0);
        int rightGain = Math.max(maxGain(node.right), 0);
        
        // Path sum including the current node and the highest gains from both sides
        int currentPathSum = node.val + leftGain + rightGain;
        
        // Update maxSum if the current path sum is greater than the previously recorded maxSum
        maxSum = Math.max(maxSum, currentPathSum);
        
        // Return the maximum gain that the current node can contribute to its parent
        return node.val + Math.max(leftGain, rightGain);
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxi = INT_MIN;
        maxPath(root, maxi);
        return maxi;
        
    }
public:
    int maxPath(TreeNode* root, int &maxi){
        if(root==NULL) return 0;
        int leftSum= max(0,maxPath(root->left, maxi));
        int rightSum= max(0,maxPath(root->right, maxi));
        maxi=max(maxi, leftSum+rightSum + root->val);
        return root->val+max(leftSum,rightSum);
    }    
};